### PR TITLE
Clean additional classes for sources passed to a compiler

### LINKED ISSUE #21644: Incremental compilation after failure issues when all classes are not removed
We got some reports from GE and also internally that after added incremental compilation after failure feature there are issues with Groovy incremental compilation. It would be good if we add an option to opt-out of this feature just in case. Or maybe even to opt-in for Groovy.

The report was, that when a super class is changed, then sometimes exception like is thrown:
```
Standard error from JVM:
    Terminating due to fatal error
    java.lang.NoClassDefFoundError: org/gradle/.../AppTest$__spock_feature_5_0_closure3
```
Note: original class name is censored since this is not public repo. Check a build scan https://e.grdev.net/s/omiun4on2i5jy to see the actual class name.

After some investigation we figured, that output directory contains wrong classes. It should contain these:
```
AppTest$__spock_feature_4_0_closure2.class
AppTest$__spock_feature_4_0_closure3$_closure15.class
AppTest$__spock_feature_4_0_closure3.class
AppTest$__spock_feature_4_1_closure4$_closure16.class
AppTest$__spock_feature_4_1_closure4.class
AppTest$__spock_feature_4_2_closure5.class
AppTest$__spock_feature_4_3_closure6.class
AppTest$__spock_feature_4_3_closure7.class
AppTest$__spock_feature_4_4_closure8.class
AppTest$__spock_feature_4_4_closure9$_closure17$_closure18.class
AppTest$__spock_feature_4_4_closure9$_closure17$_closure19.class
AppTest$__spock_feature_4_4_closure9$_closure17.class
AppTest$__spock_feature_4_4_closure9.class
AppTest$__spock_feature_4_5_closure10.class
AppTest$__spock_feature_4_5_closure11$_closure20$_closure21.class
AppTest$__spock_feature_4_5_closure11$_closure20$_closure22.class
AppTest$__spock_feature_4_5_closure11$_closure20.class
AppTest$__spock_feature_4_5_closure11.class
AppTest$__spock_feature_4_6_closure12.class
AppTest$__spock_feature_4_7_closure13.class
AppTest$__spock_feature_4_7_closure14.class
AppTest$_setup_closure1.class
AppTest.class
```

But it contained these:
```
AppTest$__spock_feature_4_1_closure4$_closure16.class
AppTest$__spock_feature_4_1_closure4.class
AppTest$__spock_feature_4_2_closure5.class
AppTest$__spock_feature_4_3_closure6.class
AppTest$__spock_feature_4_3_closure7.class
AppTest$__spock_feature_4_4_closure8.class
AppTest$__spock_feature_4_4_closure9$_closure17$_closure18.class
AppTest$__spock_feature_4_4_closure9$_closure17$_closure19.class
AppTest$__spock_feature_4_4_closure9$_closure17.class
AppTest$__spock_feature_4_4_closure9.class
AppTest$__spock_feature_4_5_closure10.class
AppTest$__spock_feature_4_5_closure11$_closure20$_closure21.class
AppTest$__spock_feature_4_5_closure11$_closure20$_closure22.class
AppTest$__spock_feature_4_5_closure11$_closure20.class
AppTest$__spock_feature_4_5_closure11.class
AppTest$__spock_feature_4_6_closure12.class
AppTest$__spock_feature_4_7_closure13.class
AppTest$__spock_feature_4_7_closure14.class
AppTest$__spock_feature_5_0_closure3$_closure15.class
AppTest$__spock_feature_5_1_closure4$_closure16.class
AppTest$__spock_feature_5_4_closure9$_closure17$_closure18.class
AppTest$__spock_feature_5_4_closure9$_closure17$_closure19.class
AppTest$__spock_feature_5_4_closure9$_closure17.class
AppTest$__spock_feature_5_5_closure11$_closure20$_closure21.class
AppTest$__spock_feature_5_5_closure11$_closure20$_closure22.class
AppTest$__spock_feature_5_5_closure11$_closure20.class
AppTest$_setup_closure1.class
AppTest.class
```

We suspect, that we don't clean all `.class` files that will be produced by the `AppTest.groovy` file before compilation. And since we don't delete some classes Groovy has them at compile time on the classpath, and that is why it creates new extra `spock_feature_5_5` classes.

While I wasn't able to reproduce the original issue,  I was able to reproduce that we don't clean all the classes for `AppTest.groovy`. 

A simpler test case that shows that we don't clean all classes is:
```
file: A.java
class A {}
```
```
file: B.java
class B extends A {}
class C {} // this class is in the same file
```

If/when we change A,  classes B and C will be recompiled since we will pass B.java to a compiler (B.class is a dependency of A.class, so we pass B.java to the compiler). But we won't delete origin C.class. So C.class will be on the classpath at the compile time.

### PR DESCRIPTION
**Pull request by asodja**
Also add a flag to opt-out of incremental compilation after failure.

Main changes:
- code that cleans additional classes in [AbstractRecompilationSpecProvider.java](https://github.com/gradle/gradle/pull/21878/files#diff-8c264c92bbb9030e5b74a3b39a56f077a72ac91d98b7bcdf005de11577ccf6ac)
- code that disables inc. compilation after a failure in [CompileTransaction.java](https://github.com/gradle/gradle/pull/21878/files#diff-9f9ba0cbef3a27aa83d8f3c7f093758db09a3a0d395d88056a5773e43dce0b72). (it's enough that we just don't revert files on exception).

Fixes #21644

### CODE REVIEW

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle test help

**Comment by asodja** [comment]
@bot-gradle test BD

**Comment by asodja** [comment]
@bot-gradle test BD

**Comment by asodja** [comment]
@bot-gradle test this PT

**Comment by asodja** [comment]
@bot-gradle test this BD

File: subprojects/docs/src/docs/userguide/jvm/groovy_plugin.adoc [hunk 1]
```diff
@@ -281,7 +281,7 @@ include::sample[dir="snippets/groovy/compilationAvoidance/kotlin/ast-transformat
 Since 5.6, Gradle introduces an experimental incremental Groovy compiler. To enable incremental compilation for Groovy, you need:
 
 * Enable <<groovy_plugin.adoc#sec:groovy_compilation_avoidance,Groovy compilation avoidance>>.
-* Explicitly enable incremental Groovy compilation in the build script:
+* Explicitly enable incremental Groovy compilation and incremental compilation after a failure in the build script:
```
Comment (nathan-contino) [inline subprojects/docs/src/docs/userguide/jvm/groovy_plugin.adoc:284]: This sentence is a little hard to read in the new state -- can we leave it as-is? Seems to me like incremental compilation includes "after a failure" by default now, so we don't need to call it out explicitly.

File: subprojects/docs/src/docs/userguide/jvm/java_plugin.adoc [hunk 2]
```diff
@@ -501,9 +501,9 @@ To help you understand how incremental compilation works, the following provides
 [[sec:incremental_compilation_known_issues]]
 === Known issues
 
-* If a compile task fails due to a compile error, it will do a full compilation again the next time it is invoked.
 * If you are using an annotation processor that reads resources (e.g. a configuration file), you need to declare those resources as an input of the compile task.
 * If a resource file is changed, Gradle will trigger a full recompilation.
+* If a custom `executable` or `javaHome` is configured on the compile task some optimizations will be deactivated, e.g. compile task will not be incremental after a compile error or if a Java constant is changed. It's recommended to use <<toolchains.adoc#toolchains, toolchains>> instead if possible.
```
Comment (nathan-contino) [inline subprojects/docs/src/docs/userguide/jvm/java_plugin.adoc:506]:
```suggestion
* Using a custom `executable` or `javaHome` deactivates some optimizations. The compile task does not use incremental build immediately after a compile error or if a Java constant changes. Use <<toolchains.adoc#toolchains, toolchains>> instead if possible.
```
Suggestion: slight rephrase.

**Review by nathan-contino — APPROVED** [review]
Docs LGTM -- minor suggestions.

File: subprojects/docs/src/docs/userguide/jvm/groovy_plugin.adoc (hunk 1 above)
Comment (asodja) [inline subprojects/docs/src/docs/userguide/jvm/groovy_plugin.adoc:284]:
```suggestion
* Explicitly enable incremental Groovy compilation in the build script:
```

Comment (asodja) [reply subprojects/docs/src/docs/userguide/jvm/groovy_plugin.adoc]: Sounds good to me, reverted that change.

File: subprojects/docs/src/docs/userguide/jvm/java_plugin.adoc (hunk 2 above)
Comment (asodja) [reply subprojects/docs/src/docs/userguide/jvm/java_plugin.adoc]: Applied suggestions

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle test this

**Review by wolfs — APPROVED** [review]
LGTM!

**Comment by wolfs** [comment]
@bot-gradle test and merge

**Comment by wolfs** [comment]
@bot-gradle test and merge

**Pull request merged**